mlir/test/Dialect/MQTOpt/Transforms/elide-permutations.mlir

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2025 Munich Quantum Software Company GmbH
+// All rights reserved.
+//
+// SPDX-License-Identifier: MIT
+//
+// Licensed under the MIT License
+
+// RUN: quantum-opt %s -split-input-file --elide-permutations | FileCheck %s
+
+// -----
+// This test checks that a single SWAP gate is removed correctly.
+
+module {
+  // CHECK-LABEL: func.func @testSingleElidePermutation
+  func.func @testSingleElidePermutation() {
+    // CHECK: %[[Q0_0:.*]] = mqtopt.allocQubit
+    // CHECK: %[[Q1_0:.*]] = mqtopt.allocQubit
+
+    // ========================== Check for operations that should be canceled ==============================
+    // CHECK-NOT: %[[ANY:.*]] = mqtopt.swap()
+
+    // ========================== Check for operations that should not be canceled ==============================
+    // CHECK: %[[Q0_1:.*]] = mqtopt.x() %[[Q0_0]]
+
+    // CHECK: mqtopt.deallocQubit %[[Q1_0]]
+    // CHECK: mqtopt.deallocQubit %[[Q0_1]]
+
+    %q0_0 = mqtopt.allocQubit
+    %q1_0 = mqtopt.allocQubit
+
+    %q0_1, %q1_1 = mqtopt.swap() %q0_0, %q1_0 : !mqtopt.Qubit, !mqtopt.Qubit
+    %q1_2 = mqtopt.x() %q1_1 : !mqtopt.Qubit
+
+    mqtopt.deallocQubit %q0_1
+    mqtopt.deallocQubit %q1_2
+
+    return
+  }
+}
+
+
+// -----
+// This test checks that a controlled SWAP gate is not removed.
+
+module {
+  // CHECK-LABEL: func.func @testControlledSwapNoElidePermutation
+  func.func @testControlledSwapNoElidePermutation() {
+    // CHECK: %[[Q0_0:.*]] = mqtopt.allocQubit
+    // CHECK: %[[Q1_0:.*]] = mqtopt.allocQubit
+    // CHECK: %[[Q2_0:.*]] = mqtopt.allocQubit
+
+    // ========================== Check for operations that should not be canceled ==============================
+    // CHECK: %[[Q01_1:.*]]:2, %[[Q2_1:.*]] = mqtopt.swap()
+
+    // CHECK: mqtopt.deallocQubit %[[Q01_1]]#0
+    // CHECK: mqtopt.deallocQubit %[[Q01_1]]#1
+    // CHECK: mqtopt.deallocQubit %[[Q2_1]]
+
+    %q0_0 = mqtopt.allocQubit
+    %q1_0 = mqtopt.allocQubit
+    %q2_0 = mqtopt.allocQubit
+
+    %q0_1, %q1_1, %q2_1 = mqtopt.swap() %q0_0, %q1_0 ctrl %q2_0 : !mqtopt.Qubit, !mqtopt.Qubit ctrl !mqtopt.Qubit
+
+    mqtopt.deallocQubit %q0_1
+    mqtopt.deallocQubit %q1_1
+    mqtopt.deallocQubit %q2_1
+
+    return
+  }
+}
+
+
+// -----
+// This test checks that all removable SWAP gates are removed.
+
+module {
+  // CHECK-LABEL: func.func @testMultiSwapElidePermutation
+  func.func @testMultiSwapElidePermutation() {
+    // CHECK: %[[Q0_0:.*]] = mqtopt.allocQubit
+    // CHECK: %[[Q1_0:.*]] = mqtopt.allocQubit
+    // CHECK: %[[Q2_0:.*]] = mqtopt.allocQubit
+
+    // ========================== Check for operations that should not be canceled ==============================
+    // CHECK-NOT: %[[ANY:.*]] = mqtopt.swap()
+
+    // ========================== Check for operations that should not be canceled ==============================
+    // CHECK: %[[ANY:.*]], %[[ANY:.*]] = mqtopt.swap() %[[ANY:.*]], %[[ANY:.*]] ctrl %[[ANY:.*]]
```

Comment:
I think it would be worth to check that the arguments of the remaining SWAP are correct instead of just asserting `ANY`.

Reply:
Ah, that's indeed a bit cumbersome. Generally, I would hope that we can at some point find better ways to write these tests to increase the iteration speed.